reflect/src/main/java/dagger/reflect/BindingGraph.java

```diff
@@ -15,105 +15,120 @@
-        : performLinking(key, binding, new LinkedHashMap<>());
+  Provider<?> getProvider(Key key) {
+    Binding binding = findBinding(key);
+    if (binding == null) {
+      throw new IllegalArgumentException("No provider available for " + key);
+    }
+
+    return binding instanceof LinkedBinding<?>
+        ? (LinkedBinding<?>) binding
+        : performLinking(key, (UnlinkedBinding) binding, new LinkedHashMap<>());
   }
 
-  private Binding<?> performLinking(Key key, Binding<?> unlinked, Map<Key, Binding<?>> chain) {
+  private LinkedBinding<?> performLinking(Key key, UnlinkedBinding unlinked,
+      Map<Key, Binding> chain) {
     chain.put(key, unlinked);
 
-    Key[] dependencyKeys = unlinked.dependencies();
-    Binding<?>[] dependencies = new Binding[dependencyKeys.length];
-
-    for (int i = 0; i < dependencyKeys.length; i++) {
-      Key dependencyKey = dependencyKeys[i];
-      @Nullable Binding<?> dependency = locateBinding(dependencyKey);
-
-      if (dependency == null && Types.equals(Types.getRawType(key.type()), Optional.class)) {
-        continue;
-      }
-
-      if (!dependency.isLinked()) {
-        if (chain.containsKey(dependencyKey)) {
+    Binding.LinkRequest request = unlinked.request();
+    Key[] requestKeys = request.keys;
+    boolean[] requestOptionals = request.optionals;
+
+    LinkedBinding<?>[] dependencies = new LinkedBinding<?>[requestKeys.length];
+    for (int i = 0; i < requestKeys.length; i++) {
+      Key requestKey = requestKeys[i];
+      boolean isOptional = requestOptionals[i];
+      Binding dependency = findBinding(requestKey);
+
+      LinkedBinding<?> linkedBinding;
+      if (dependency instanceof LinkedBinding<?>) {
+        linkedBinding = (LinkedBinding<?>) dependency;
+      } else if (dependency instanceof UnlinkedBinding) {
+        if (chain.containsKey(requestKey)) {
           StringBuilder builder = new StringBuilder("Dependency cycle detected!\n");
-          for (Map.Entry<Key, Binding<?>> entry : chain.entrySet()) {
+          for (Map.Entry<Key, Binding> entry : chain.entrySet()) {
             builder.append(" * Requested: ")
                 .append(entry.getKey())
                 .append("\n     from ")
                 .append(entry.getValue())
                 .append('\n');
           }
           builder.append(" * Requested: ")
-              .append(dependencyKey)
+              .append(requestKey)
               .append("\n     which forms a cycle.");
           throw new IllegalStateException(builder.toString());
         }
-
-        dependency = performLinking(dependencyKey, dependency, chain);
+        linkedBinding = performLinking(requestKey, (UnlinkedBinding) dependency, chain);
+      } else if (isOptional) {
+        linkedBinding = null;
+      } else {
+        StringBuilder builder = new StringBuilder("Cannot locate binding for ");
+        builder.append(requestKey).append('\n');
+        for (Map.Entry<Key, Binding> entry : chain.entrySet()) {
```

Comment:
extract this to a method and reuse above as well? something like:
here:
`appendRequests(builder, chain, requestKey, "was not found");`
and above:
`appendRequests(builder, chain, requestKey, "forms a cycle");`

it would reduce complexity and dedupe, + the method would fit on one 1080p screen :)

Reply:
I'm probably going to extract both to individual static methods for creating the exception so it can be re-used for `getProvider`. For a follow-up, though.